torch/csrc/utils/tensor_numpy.cpp

```diff
@@ -386,6 +396,314 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
       at::device(kCUDA).dtype(dtype)
   );
 }
+
+
+template <typename T>
+std::string get_vector_str(std::vector<T> & vec, int64_t start, int64_t end=-1) {
+  /*
+  A utility function for returning a string representation of a std::vector containing ints/floats.
+  Useful for printing the shape of an N-dimensional tensor. This is used in the error message creation
+  part in the _extract_ndarrays function below.
+  */
+  std::ostringstream output;
+  output << "(";
+
+  if (end < 0) {
+    end = vec.size();
+  }
+  for (int64_t i = start; i < end; ++i) {
+    if (i != end - 1) {
+      output << vec[i] << ", ";        
+    } else {
+      output << vec[i] << ")";
+    }
+  }
+  return output.str();
+}  
+
+template <typename T>
+std::string get_buffer_str(void* buffer, int64_t start, int64_t end) {
+  /*
+  A similar utility function as get_vector_str, except it works on an array buffer instead of vector.
+  */
+  T* array = static_cast<T*>(buffer); 
+  std::ostringstream output;
+  output << "(";
+
+  for (int64_t i = start; i < end; ++i) {
+    if (i != end - 1) {
+      output << array[i] << ", ";        
+    } else {
+      output << array[i] << ")";
+    }
+  }
+  return output.str();
+}  
+
+inline bool _shape_match_ndarray(std::vector<int64_t> & shape, PyArrayObject* array, int64_t start) {
+  /*
+  A utility function that returns true only if shape[start:] == array.shape where array is a numpy array.
+  Else returns false. This is used in _extract_ndarrays() below for checking whether an embedded numpy array
+  has the target shape.
+  */
+  int64_t ndim = PyArray_NDIM(array);
+  if ((shape.size() - start) != ndim) {
+    return false;
+  }
+
+  npy_intp* array_shape = PyArray_DIMS(array);
+  for (int64_t i = 0; i < ndim; ++i) {
+    if (array_shape[i] != shape[i+start]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+
+inline int _validate_ndarray(PyObject* obj, std::vector<int64_t> & shape, std::vector<int64_t> & ndarray_shape, int64_t depth) {
+  /*
+  This returns -1 if obj is a scalar array, 0 if obj is not array else if 1 given no exception is raised.
+  A ValueError will be raised if the depth of obj within the input iterable, as given to _extract_ndarrays,
+  is not the same as any other array within the input iterable OR if its shape does not match any other
+  array's shape in the iterable.
+  */
+
+  PyArrayObject* array;
+  if (PyArray_Check(obj)) {
+    array = (PyArrayObject*) obj;
+  } else {
+    return 0;
+  } 
+
+  if (PyArray_NDIM(array) == 0) {
+    return -1;
+  }
+
+  if (ndarray_shape.size() == 0) {
+    /*
+    If ndarray_shape is an empty vector, then ndarray_shape[0] is set to the depth of the array in the iterable and
```

Comment:
This reuse of one data structure for two things is a little confusing for the reader. 

I'm a little confused why the algorithm doesn't work conceptually like this:

- recurse down leftmost branches of the forest until it finds an array, identifies the (common) depth and (common) shape of that array, and then returns those values (either by modifying inputs or returning a tuple)
- then discover each array in the forest and validate it has the properties of the first array

The current algorithm may be actually be doing something similar to this, but it's tricky to read.

Reply:
```_validate_ndarray``` first finds out the depth and shape of the first numpy array it sees and then in each subsequent calls it validates whether the rest of the arrays have same shape and depth. So yes the current algorithm does that using a single function. Doing this using a single function is more concise than breaking it apart. I have replaced the single datastructure with a pair where pair.first is the depth and pair.second is the vector that represents the target shape.